kube-test/src/k3d.rs

```diff
@@ -0,0 +1,222 @@
+impl TestEnvBuilder {
+    /// Set the number of servers in the temporary cluster.
+    pub fn servers(&mut self, servers: usize) -> &mut Self {
+        self.servers = servers;
+        self
+    }
+
+    /// Set the number of agents in the temporary cluster.
+    pub fn agents(&mut self, agents: usize) -> &mut Self {
+        self.agents = agents;
+        self
+    }
+
+    /// Enable host ip injection.
+    pub fn inject_host_ip(&mut self) -> &mut Self {
+        self.host_ip_injection = true;
+        self
+    }
+
+    /// Create image volume.
+    pub fn with_image_volume(&mut self) -> &mut Self {
+        self.create_image_volume = true;
+        self
+    }
+
+    /// Create load balancer.
+    pub fn with_load_balancer(&mut self) -> &mut Self {
+        self.create_load_balancer = true;
+        self
+    }
+
+    /// Set `verbose` flag.
+    pub fn verbose(&mut self) -> &mut Self {
+        self.verbose = true;
+        self
+    }
+
+    /// Set the k3s version to use. The version should be in the form `v1.20.6`.
+    pub fn k3s_version<T: Into<String>>(&mut self, version: T) -> &mut Self {
+        self.version = Some(version.into());
+        self
+    }
+
+    /// Create the test environment.
+    pub fn build(&self) -> TestEnv {
+        let name = xid::new().to_string();
+        let servers = format!("--servers={}", self.servers);
+        let agents = format!("--agents={}", self.agents);
+        let mut args = vec![
+            "cluster",
+            "create",
+            &name,
+            "--wait",
+            // Don't change `~/.kube/config`
+            "--kubeconfig-update-default=false",
+            "--kubeconfig-switch-context=false",
+            // Disable to avoid having to create the default service account in each test.
+            // > we're keeping this disabled because if enabled, default SA is
+            // > missing which would force all tests to create one
+            // > in normal apiserver operation this SA is created by controller, but that is
+            // > not run in integration environment
+            // > https://git.io/JZKFC
+            "--k3s-server-arg",
+            "--kube-apiserver-arg=disable-admission-plugins=ServiceAccount",
+            // Disable components and features
+            "--k3s-server-arg",
+            "--disable=servicelb",
+            "--k3s-server-arg",
+            "--disable=traefik",
+            "--k3s-server-arg",
+            "--disable=metrics-server",
+            "--k3s-server-arg",
+            "--disable-cloud-controller",
+            "--no-rollback",
+            &servers,
+            &agents,
+        ];
+        if self.verbose {
+            args.push("--verbose");
+        }
+        if !self.host_ip_injection {
+            args.push("--no-hostip");
+        }
+        if !self.create_image_volume {
+            args.push("--no-image-volume");
+        }
+        if !self.create_load_balancer {
+            args.push("--no-lb");
+        }
```

Comment:
I love this. The default here being to make a k3d as small as possible.

Reply:
It does set a fairly weird precedent for the future though, where new K3d features will start out as force-enabled, and then switch to disabled-by-default as kube-rs learns about it. (Though tbf I'd put this more on K3d than on kube-rs..)